### PR TITLE
First attempt at softmax learner for cbadf

### PR DESCRIPTION
**Pull request by adith387**
- Do scores from cb-adf.predict have to be negated before taking softmax? (i.e. does rest of vw treat lower scores as better?)
- Is it correct to treat cb-adf.predict as the current regressor's scores?
- Does DM handle shared features properly? Current implementation builds on DM since it is conceptually simpler than MTR.
- When creating $k$ squared-loss-regression examples, the original multi-line example weights are backed up and restored. Do we also need to back-up and restore the original number of features in each example?

### CODE REVIEW

**Comment by JohnLangford** [comment]
@adith387 status on this?

**Comment by JohnLangford** [comment]
Ping?

**Comment by marco-rossi29** [comment]
I've run `--cb_type sm` with a sample dataset from cmplx, and I see an issue in the predictions (see `current predict` column) where the probability is greater than 1:
```
>vw --dsjson D:\data\logDownloader_v5\cmplx\cmplx_deepmerged_data_2018-10-01_2018-10-14.json.gz --cb_type sm --cb_adf -P 50000 -c
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using cache_file = D:\data\logDownloader_v5\cmplx\cmplx_deepmerged_data_2018-10-01_2018-10-14.json.gz.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
-0.013296 -0.013296        50000        50000.0    known        4:10.6072...       73
-0.013022 -0.012749       100000       100000.0    known        0:14.4278...       54
-0.013974 -0.015878       150000       150000.0    known        0:17.1166...       69
-0.014888 -0.017628       200000       200000.0    known        0:17.6764...       57
-0.013458 -0.007738       250000       250000.0    known        0:22.8696...      156
-0.012211 -0.005979       300000       300000.0    known        0:12.4104...       90
-0.011255 -0.005517       350000       350000.0    known        2:25.4138...       56
-0.010141 -0.002342       400000       400000.0    known        0:24.7752...       93
-0.009829 -0.007330       450000       450000.0    known        2:33.932...       52
-0.009511 -0.006655       500000       500000.0    known        3:30.5777...       52
-0.009322 -0.007425       550000       550000.0    known        0:30.3612...       77
-0.008964 -0.005034       600000       600000.0    known        1:27.9163...       99

finished run
number of examples = 613086
weighted example sum = 613086.000000
weighted label sum = 0.000000
average loss = -0.008962
total feature number = 42089178
```
Using `mtr` gives predictions with a real pdf:
```
>vw --dsjson D:\data\logDownloader_v5\cmplx\cmplx_deepmerged_data_2018-10-01_2018-10-14.json.gz --cb_type mtr --cb_adf -P 50000 -c
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using cache_file = D:\data\logDownloader_v5\cmplx\cmplx_deepmerged_data_2018-10-01_2018-10-14.json.gz.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
-0.016839 -0.016839        50000        50000.0    known        1:0.208558...       73
-0.015842 -0.014845       100000       100000.0    known        1:0.0445442...       54
-0.015125 -0.013690       150000       150000.0    known        5:0.0301438...       69
-0.015519 -0.016703       200000       200000.0    known        1:0.133496...       57
-0.014740 -0.011621       250000       250000.0    known        2:0.164838...      156
-0.014324 -0.012245       300000       300000.0    known        0:0.0599655...       90
-0.014063 -0.012497       350000       350000.0    known        0:0.0595418...       56
-0.013963 -0.013264       400000       400000.0    known        1:0.12498...       93
-0.014031 -0.014572       450000       450000.0    known        2:0.0144068...       52
-0.013754 -0.011267       500000       500000.0    known        2:0.0596526...       52
-0.013834 -0.014630       550000       550000.0    known        1:-0.0112924...       77
-0.014027 -0.016153       600000       600000.0    known        2:-0.0609681...       99

finished run
number of examples = 613086
weighted example sum = 613086.000000
weighted label sum = 0.000000
average loss = -0.013905
total feature number = 42089178
```

**Comment by JohnLangford** [comment]
I suspect at least some of this is a reporting issue rather than a learning dynamics issue.

File: vowpalwabbit/cb_adf.cc [hunk 1]
```diff
@@ -83,6 +88,110 @@ void learn_IPS(cb_adf& mydata, multi_learner& base, multi_ex& examples)
   call_cs_ldf<true>(base, examples, mydata.cb_labels, mydata.cs_labels, mydata.prepped_cs_labels, mydata.offset);
 }
 
+float safe_probability(float prob)
+{
+  if (prob <= 0.)
+  {
+    std::cout << "Probability " << prob << " is not possible, replacing with 1e-6.  Fix your dataset. " << std::endl;
```
Comment (jackgerrits) [inline vowpalwabbit/cb_adf.cc]: Error messages or warnings should be printed to cerr

**Comment by JohnLangford** [comment]
The default in VW has been that diagnostics are on stderr and error
messages on stdout.

-John

**Comment by jackgerrits** [comment]
Doesn't all.trace_message go to stderr though?

**Comment by JohnLangford** [comment]
I don't recall.

Using all.trace_message is better anyways since it can work for a library,
unlike a hard-coded choice.

-John

File: vowpalwabbit/cb_adf.cc (hunk 1 above)
Comment (adith387) [reply vowpalwabbit/cb_adf.cc]: Made the following changes:
cb_adf.cc: Removed the safe_probability method. Replaced all calls to it to call GEN_CS::safe_probability instead. Having 2 safe_probability methods is not healthy. 
gen_cs_example.cc::safe_probability prints error message to std::cout and needs to be fixed in the master branch.

**Comment by jackgerrits** [comment]
@adith387 just letting you know I fixed the merge conflict so you'll need to pull before submitting any more commits.

**Comment by jackgerrits** [comment]
@adith387 we need a test in RunTests for this, do you need a hand with that?

**Comment by adith387** [comment]
> @adith387 we need a test in RunTests for this, do you need a hand with that?

Yes please!

**Comment by jackgerrits** [comment]
Okay no problem, I can actually just check in a test for you. Would you be able to share me the hand crafted data set you made?

**Comment by jackgerrits** [comment]
I was unable to commit to this branch so I've created a PR on your fork adding the test: https://github.com/adith387/vowpal_wabbit/pull/1

**Comment by jackgerrits** [comment]
Thanks @adith387 ], merged!
